### ISSUE TITLE
rust-analyzer panics if a format string refers to a nonexitent variable.

### ISSUE THREAD

**Issue by TDecking** [rust-lang/rust-analyzer#16682]
**rust-analyzer version**: rust-analyzer version: 0.3.1860-standalone (5346002d0 2024-02-25)

**rustc version**: rustc 1.77.0-nightly (bf8716f1c 2023-12-24)

**relevant settings**: None

**code snippet to reproduce**: 
```rust
fn main() {
    //println!("Hello, world!{va}");
}
```

https://github.com/rust-lang/rust-analyzer/assets/15847799/eca2391f-c545-42ea-9a7f-8fafb89a7f52

**log output**:
```text
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'LspServer' panicked at D:\a\rust-analyzer\rust-analyzer\crates\stdx\src\thread\pool.rs:86:35:
called `Result::unwrap()` on an `Err` value: "SendError(..)"
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: rust_analyzer::task_pool::TaskPool<T>::spawn
   4: rust_analyzer::dispatch::RequestDispatcher::on_with_thread_intent
   5: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::prime_caches::{{closure}}::{{closure}}
   6: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::run
   7: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::run
   8: rust_analyzer::main_loop::main_loop
   9: rust_analyzer::run_server
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread '<unnamed>' panicked at C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\lsp-server-0.7.6\src\stdio.rs:28:37:
receiver was dropped, failed to send a message: "SendError(..)"
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\jod-thread-0.1.2\src\lib.rs:33:22:
called `Result::unwrap()` on an `Err` value: Any { .. }
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: stdx::thread::JoinHandle<T>::join
   4: lsp_types::inline_completion::_::<impl serde::ser::Serialize for lsp_types::inline_completion::InlineCompletionOptions>::serialize
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[Info  - 16:33:28] Connection to server got closed. Server will restart.
true
```

**Comment by TDecking** [comment]
I also decided to test this after running `rustup update`. The crash persists.

**Comment by Veykril** [comment]
Fixed in  #16684
